module/os/linux/spl/spl-zlib.c

```diff
@@ -202,7 +202,7 @@ spl_zlib_init(void)
 	zlib_workspace_cache = kmem_cache_create(
 	    "spl_zlib_workspace_cache",
 	    size, 0, NULL, NULL, NULL, NULL, NULL,
-	    KMC_VMEM);
+	    KMC_KVMEM);
```

Comment:
I think that after this change there are no kmem caches that use KMC_VMEM.  So the only use of KM_VMEM is from vmem_alloc().

I'm trying to understand when we use kmem_alloc() vs vmem_alloc().  (On illumos, ZFS doesn't use vmem_alloc.)  It looks like we use vmem_alloc() when doing an allocation that might be > spl_kmem_alloc_max, since in this case kmem_alloc() would fail.  Could we instead always use kmem_alloc(), but have kmem_alloc try kmalloc first and then fall back on vmalloc?  I.e. use kvmalloc on kernels that have it, and write the fallback ourself for older kernels?

Reply:
Meh. Mail is evil.

well, this is what the whole change is about. Use kvmalloc when it‘s there, else use our own implementation

Reply:
I saw `spl_kvmalloc()` in `spl-kmem.c`, but I also see a bunch of calls to `spl_vmem_alloc()`, which calls `spl_kmem_alloc_impl(KM_VMEM)` which calls `__vmalloc()` (if `size > spl_kmem_alloc_max`) instead of `spk_kvmalloc()`.

My question is: could `spl_kmem_alloc_impl()` always call `spl_kvmalloc()`, regardless of `size` and `flags`

Reply:
@ahrens. basicly yes. (except for some special purpose memory like dma related transfers, but that shouldnt matter in a filesystem driver) thats the intention of the patch. but its not the purpose of the patch to convert everything right now. its now more important to get it merged. based on it future improvements can be made in the rest of zfs. the kmem_alloc_max restrictions is bad at all. i see no reason why this is neccessary. if kmem is available at that size there is no reason to avoid allocation. i dont know why someone wrote this rule in the spl code

Reply:
OK so the goal is to do what I suggested and this PR is the first step.  Thank you.

Reply:
That's right.  This is intended to be just a first step.